apps/condo/domains/property/components/panels/Builder/MapConstructor.ts

```diff
@@ -0,0 +1,490 @@
+                type: 'floor',
+                units,
+            })
+        }
+        this.map.sections.push(newSection)
+        this.visibleSections.push(newSection.id)
+        this.updateFormField()
+    }
+    
+    public mapUpdateSection (section: BuildingSection): void {
+        const sectionIndex = this.map.sections.findIndex(mapSection => section.id === mapSection.id)
+        if (sectionIndex !== -1) {
+            this.map.sections[sectionIndex].name = section.name
+        }
+        this.editMode = 'addSection'
+        this.updateFormField()
+    }
+    
+    public mapRemoveSection (id: string): void {
+        const sectionIndex = this.map.sections.findIndex(mapSection => mapSection.id === id)
+        this.map.sections.splice(sectionIndex, 1)
+        this.editMode = 'addSection'
+        this.updateFormField()
+    }
+
+    public mapAddUnit (unit: BuildingUnit): void {
+        const { id, section, floor, label } = unit
+        const sectionIndex = this.map.sections.findIndex(mapSection => mapSection.id === section)
+        if (sectionIndex === -1) {
+            return
+        }
+        const floorIndex = this.map.sections[sectionIndex].floors.findIndex(sectionFloor => sectionFloor.id === floor)
+        if (floorIndex === -1) {
+            return
+        }
+        const newUnit = {
+            id,
+            label, 
+            type: 'unit',
+        }
+        if (!id) {
+            newUnit.id = String(++this.autoincrement)
+        }
+        this.map.sections[sectionIndex].floors[floorIndex].units.push(newUnit)
+        this.updateUnitNumbers(newUnit)
+        this.editMode = 'addUnit'
+        this.updateFormField()
+    }
+
+    public mapRemoveUnit (id: string): void {
+        const unitIndex = this.getUnitIndex(id)
+        if (unitIndex.unit !== -1) {
+            this.map.sections[unitIndex.section].floors[unitIndex.floor].units.splice(unitIndex.unit, 1)
+        }
+        this.selectedUnit = null
+        this.editMode = 'addUnit'
+        this.updateFormField()
+    }
+
+    public mapUpdateUnit (unit: BuildingUnit): void {
+        const unitIndex = this.getUnitIndex(unit.id)
+        if (unitIndex.unit === -1) {
+            return
+        }
+        const oldSection = this.map.sections[unitIndex.section].id
+        const oldFloor = this.map.sections[unitIndex.section].floors[unitIndex.floor].id
+        
+        if (oldFloor !== unit.floor || oldSection !== unit.section) {
+            this.mapRemoveUnit(unit.id)
+            this.mapAddUnit(unit)
+        } else {
+            this.map.sections[unitIndex.section].floors[unitIndex.floor].units[unitIndex.unit].label = unit.label
+            this.updateUnitNumbers(unit)
+        }
+        this.editMode = 'editUnit'
+        this.updateFormField()
+    }
+
+    private updateFormField () {
+        this.updateMap(this.map)
+    }
+    private updateMap: (map: BuildingMap) => void
+}
+
+
+class MapDebug extends MapEdit {
```

Comment:
Do we really need it?

Reply:
Removed - I use it in tests to have access to JSON. But I think there is no need to protect JSON - as we have validation on save